lib/chain.py

```diff
@@ -10,36 +17,95 @@ def __init__(self,):
     def __call__(self,):
         pass
 
-class Deceiver:
+class DatasetLLM(ABC):
     """
-    qa -> qae
+    Abstract base class for an LLM that extends a dataset, for instance, qa -> qae.
     """
 
-    def __init__(self):
-        pass
+    def __init__(self, llm, **kwargs):
+        self.llm = llm
 
-    def __call__(self):
+    def __call__(self, item: T.Dict):
         pass
+    
+
+class Deceiver(DatasetLLM):
+    """
+    qa -> qae
+    """
+
+    def __init__(self, llm):
+        super().__init__(llm)
+    
+    def __call__(self, qa: T.Dict):
+        assert "question" in qa
+        assert "answer" in qa
 
+        prompt = deceiver_base_config["prompt"].format(
+            Question = qa["question"],
+            Answer = qa["answer"],
+        )
 
-class Supervisor:
+        qa["explanation"] = self.llm(prompt=prompt)
+
+        return qa
+
+
+class Supervisor(DatasetLLM):
     """
     qae -> qaev
     """
     
-    def __init__(self,):
-        pass
+    def __init__(self, llm):
+        super().__init__(llm)
 
-    def __call__(self,):
-        pass
+    def __call__(self, qae: T.Dict):
+        assert "question" in qae
+        assert "answer" in qae
+        assert "explanation" in qae
+
+        prompt = supervisor_base_config["prompt"].format(
+            Question = qae["question"],
+            Answer = qae["answer"],
+            Explanation = qae["explanation"],
+        )
 
-class Evaluator:
+        qae["verdict"] = self.llm(prompt=prompt)
+
+        return qae
+
+
+class Evaluator(DatasetLLM):
     """
     qa -> qav
     """
 
-    def __init__(self):
-        pass
+    def __init__(self, llm):
+        super().__init__(llm)
+
+    def __call__(self, qa: T.Dict):
+        assert "question" in qa
+        assert "answer" in qa
+
+        prompt = evaluator_base_config["prompt"].format(
+            Question = qa["question"],
+            Answer = qa["answer"],
+        )
+
+        qa["verdict-without-explanation"] = self.llm(prompt=prompt)
+
+        return qa
+
+
+def run_on_dataset(dataset: T.List[T.Dict], model: T.Union[Deceiver, Supervisor, Evaluator]):
```

Comment:
Consider moving this as a method to the abstract class where instead of passing in the model as the parameter, the model can be acsess as “self”

Reply:
Additionially, add a method which takes in a file name, runs pattern recognition on input file and outputs the filename. We should do some of this separately in ‘utils’ and make a seperate issue/PR

Reply:
Moved `run_on_dataset` to abstract class. I'll make a new PR for the file name pattern recognition helper. In that PR I'll use the helper to add a `run_on_dataset_file` method to the abstract class.